src/tpm/mod.rs

```diff
@@ -0,0 +1,158 @@
+mod tpm_wrapper;
+
+use std::{
+    convert::{TryFrom, TryInto},
+    sync::Once,
+};
+use std::sync::{Arc, Mutex};
+use thiserror::Error;
+use p256::{ecdsa};
+use p256::elliptic_curve::sec1::FromEncodedPoint;
+use sha2::{Digest, Sha256};
+
+use crate::{keypair, KeyTag, Network, public_key, Result, KeyType as CrateKeyType,
+            error, ecc_compact, ecc_compact::Signature};
+
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("function {0} returned error code {1}")]
+    TPMError(String, u32),
+
+    #[error("wrong key path")]
+    WrongKeyPath,
+}
+
+impl Error {
+    pub fn tpm_error(func: String, code: u32) -> crate::Error {
+        Self::TPMError(func, code).into()
+    }
+
+    pub fn wrong_key_path() -> crate::Error {
+        Self::WrongKeyPath.into()
+    }
+}
+
+
+static INIT: Once = Once::new();
```

Comment:
My gut tells me that both INIT and COUNTER can be replaced with a single global object that eventually calls Fapi_Finalize when the program exits via Drop. The specifics depend on whether or not Fapi calls are thread safe or not (I have no idea).

But I think at a minimum `COUNTER` can converted to a `std::sync::atomic::AtomicI32`.